File: DSA/src/Questions/DP/SubsetSum.java
package Questions.DP;

import java.util.Arrays;

public class SubsetSum {
    // Problem:- We are given an array ‘ARR’ with N positive integers. We need to
    // find if there is a subset in “ARR” with a sum equal to K. If there is, return
    // true else return false

    // Helper function to solve subset sum problem using dynamic programming
    static boolean subsetSumUtil(int ind, int target, int[] arr, int[][] dp) {
        // If the target sum is achieved, return true
        if (target == 0)
            return true;

        // If we have considered all elements but haven't reached the target, return
        // false
        if (ind == 0)
            return arr[0] == target;

        // If the result for this subproblem has already been calculated, return it
        if (dp[ind][target] != -1)
            return dp[ind][target] != 0;

        // Try not taking the current element
        boolean notTaken = subsetSumUtil(ind - 1, target, arr, dp);

        // Try taking the current element if it doesn't exceed the target
        boolean taken = false;
        if (arr[ind] <= target)
            taken = subsetSumUtil(ind - 1, target - arr[ind], arr, dp);

        // Store the result in the DP table and return whether either option was
        // successful
        dp[ind][target] = notTaken || taken ? 1 : 0;
        return notTaken || taken;
    }

    // Main function to check if there exists a subset with a given target sum
    static boolean subsetSumToKMemorization(int n, int k, int[] arr) {
        // Create a DP table with dimensions [n][k+1]
        int dp[][] = new int[n][k + 1];

        // Initialize DP table with -1 (unprocessed)
        for (int row[] : dp)
            Arrays.fill(row, -1);

        // Call the recursive helper function
        return subsetSumUtil(n - 1, k, arr, dp);
    }

    // Function to check if there exists a subset with a given target sum
    static boolean subsetSumToKTabulation(int n, int k, int[] arr) {
        // Each (x,y) cell indicates that with the first x elements, we can achieve y
        boolean dp[][] = new boolean[n][k + 1];

        // 0 sum can be achieved always
        for (int i = 0; i < n; i++) {
            dp[i][0] = true;
        }

        // With the first element, we can achieve the sum equal to the value of the
        // first
        if (arr[0] <= k) {
            dp[0][arr[0]] = true;
        }

        // Fill in the DP table using bottom-up approach
        for (int ind = 1; ind < n; ind++) {
            for (int target = 1; target <= k; target++) {
                // Calculate if the current target can be achieved without taking the current
                // element
                boolean notTaken = dp[ind - 1][target];

                // Calculate if the current target can be achieved by taking the current element
                boolean taken = false;
                if (arr[ind] <= target) {
                    // If the current element is less than or equal to the target, then we can
                    // finding the target - arr[ind] elements state using the previous elements
                    taken = dp[ind - 1][target - arr[ind]];
                }

                // Store the result in the DP table
                dp[ind][target] = notTaken || taken;
            }
        }

        // The final result is stored in the bottom-right cell of the DP table
        return dp[n - 1][k];
    }

    // Function to check if there exists a subset with a given target sum
    static boolean subsetSumToKOptimized(int n, int k, int[] arr) {
        // Create an array to store the previous row of the DP table
        boolean prev[] = new boolean[k + 1];

        // Initialize the first row of the DP table
        prev[0] = true;

        // Initialize the first column of the DP table
        if (arr[0] <= k) {
            prev[arr[0]] = true;
        }

        // Fill in the DP table using bottom-up approach
        for (int ind = 1; ind < n; ind++) {
            // Create an array to store the current row of the DP table
            boolean cur[] = new boolean[k + 1];

            // Initialize the first column of the current row
            cur[0] = true;

            for (int target = 1; target <= k; target++) {
                // Calculate if the current target can be achieved without taking the current
                // element
                boolean notTaken = prev[target];

                // Calculate if the current target can be achieved by taking the current element
                boolean taken = false;
                if (arr[ind] <= target) {
                    taken = prev[target - arr[ind]];
                }

                // Store the result in the current row of the DP table
                cur[target] = notTaken || taken;
            }

            // Update the previous row with the current row
            prev = cur;
        }

        // The final result is stored in the last cell of the previous row
        return prev[k];
    }

    public static void main(String[] args) {
        int arr[] = { 3, 34, 4, 12, 5, 2 };
        int n = arr.length;
        int k = 9;

        System.out.println(subsetSumToKMemorization(n, k, arr));
        System.out.println(subsetSumToKTabulation(n, k, arr));
        System.out.println(subsetSumToKOptimized(n, k, arr));
    }
}

File: DSA/Python/Questions/DP/subset_sum.py
"""
# Question: Subset Sum Problem
# Link: https://leetcode.com/problems/partition-equal-subset-sum/

# Problem Statement:
# Given an array 'ARR' with N positive integers. Find if there is a subset in ARR
# with a sum equal to K. Return true if such a subset exists, false otherwise.

# Example:
# Input: arr = [1, 2, 3, 4], k = 7
# Output: True
# Explanation: Subset [3, 4] sums to 7
"""

from typing import List

# 1. Recursive Solution with Memoization
"""
Algorithm:
1. For each element, we have two choices:
   - Include in sum
   - Exclude from sum
2. Track current index and target sum
3. Use dp table to memoize results

Time Complexity: O(N*K) where N is array length and K is target sum
Space Complexity: O(N*K) + O(N) for dp array and recursion stack
"""


def subset_sum_recursive(arr: List[int], k: int) -> bool:
    n = len(arr)
    dp = [[-1] * (k + 1) for _ in range(n)]

    def solve(ind: int, target: int) -> bool:
        # Base case: target achieved
        if target == 0:
            return True

        # Base case: first element
        if ind == 0:
            return arr[0] == target

        # Return memoized result
        if dp[ind][target] != -1:
            return dp[ind][target] == 1

        # Try not taking current element
        not_take = solve(ind - 1, target)

        # Try taking current element if possible
        take = False
        if arr[ind] <= target:
            take = solve(ind - 1, target - arr[ind])

        # Store result in dp
        dp[ind][target] = 1 if (take or not_take) else 0
        return take or not_take

    return solve(n - 1, k)


# 2. Tabulation Solution
"""
Algorithm:
1. Create dp table of size n * (k+1)
2. Initialize base cases for first row
3. For each element and sum:
   - Try including and excluding current element
   - Store result in dp table

Time Complexity: O(N*K)
Space Complexity: O(N*K)
"""


def subset_sum_tabulation(arr: List[int], k: int) -> bool:
    n = len(arr)
    dp = [[False] * (k + 1) for _ in range(n)]

    # Initialize first row
    for i in range(n):
        dp[i][0] = True
    if arr[0] <= k:
        dp[0][arr[0]] = True

    # Fill dp table
    for ind in range(1, n):
        for target in range(1, k + 1):
            not_take = dp[ind - 1][target]

            take = False
            if arr[ind] <= target:
                take = dp[ind - 1][target - arr[ind]]

            dp[ind][target] = take or not_take

    return dp[n - 1][k]


# 3. Space Optimized Solution
"""
Algorithm:
1. Use single array instead of 2D array
2. Update array for each element
3. Process right to left to avoid overwriting

Time Complexity: O(N*K)
Space Complexity: O(K)
"""


def subset_sum_optimized(arr: List[int], k: int) -> bool:
    n = len(arr)
    prev = [False] * (k + 1)

    # Initialize first row
    prev[0] = True
    if arr[0] <= k:
        prev[arr[0]] = True

    # Process each element
    for ind in range(1, n):
        curr = [False] * (k + 1)
        curr[0] = True
        for target in range(1, k + 1):
            not_take = prev[target]

            take = False
            if arr[ind] <= target:
                take = prev[target - arr[ind]]

            curr[target] = take or not_take
        prev = curr[:]

    return prev[k]


def main():
    # Test cases
    test_cases = [
        {"arr": [1, 2, 3, 4], "k": 7},
        {"arr": [2, 3, 7, 8, 10], "k": 11},
        {"arr": [1, 2, 3], "k": 7},
        {"arr": [3], "k": 3},
    ]

    for i, test in enumerate(test_cases):
        print(f"\nTest Case {i + 1}:")
        print(f"Array: {test['arr']}")
        print(f"Target Sum: {test['k']}")
        print(f"Recursive Solution: {subset_sum_recursive(test['arr'], test['k'])}")
        print(f"Tabulation Solution: {subset_sum_tabulation(test['arr'], test['k'])}")
        print(f"Optimized Solution: {subset_sum_optimized(test['arr'], test['k'])}")


if __name__ == "__main__":
    main()
